src/server/proxy_protocol.rs

```diff
@@ -127,10 +157,35 @@ where
     });
 }
 
-/// Constructs a hash key based on the source ip and the destination port
-/// in a straightforward consistent way
-pub fn construct_proxy_hash_key(source: &IpAddr, port: u16) -> String {
-    format!("{}.{}", source, port)
+/// The key is constructed out of the external source IP of the client and the passive listening port that has been reserved
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct ProxyHashKey {
+    source: IpAddr,
+    port: u16,
+}
+
+impl ProxyHashKey {
+    fn new(source: IpAddr, port: u16) -> Self {
+        ProxyHashKey { source, port }
+    }
+}
+
+impl From<&ProxyConnection> for ProxyHashKey {
+    fn from(connection: &ProxyConnection) -> Self {
+        ProxyHashKey::new(connection.source.ip(), connection.destination.port())
+    }
+}
+
+impl ToString for ProxyHashKey {
+    fn to_string(&self) -> String {
+        let s = self.source.to_string();
+        let p = self.port.to_string();
+        let mut result = String::with_capacity(s.len() + p.len() + 1);
+        result.push_str(&s);
+        result.push('.');
+        result.push_str(&p);
+        result
```

Comment:
is there a reason to not just `format!("{s}.{p}")`?

Reply:
I was just playing with avoiding memory to be copied unnecessarily. I know it's not making much difference.